File: bcs-runtime/bcs-k8s/bcs-component/bcs-k8s-driver/kubedriver/versions/versions.go
/*
 * Tencent is pleased to support the open source community by making Blueking Container Service available.
 * Copyright (C) 2019 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * http://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package versions

import (
	"errors"
	"fmt"
	"io/ioutil"
	"path/filepath"
	"regexp"
	"strings"

	jsoniter "github.com/json-iterator/go"
)

//ClientSetter client set for multiple version
type ClientSetter struct {
	ClientSet   string
	BodyContent *[]byte
}

var json = jsoniter.ConfigCompatibleWithStandardLibrary

//IfWithClientSet check api prefix
func (cs *ClientSetter) IfWithClientSet(uri string) bool {
	if strings.HasPrefix(uri, "api") || strings.HasPrefix(uri, "apis") {
		return true
	}
	return false
}

//GetClientSetUrl get uri relative version + group
func (cs *ClientSetter) GetClientSetUrl(uri string, version string, apiPrefer map[string]string) error {
	formattedUri := FormatURI(uri)
	allUrl := apiVersionMap[version]

	for _, url := range allUrl {
		formerUrl, laterUrl, err := SplitUrlByVersion(url)

		if err != nil {
			continue
		}
		if laterUrl == formattedUri {
			formerUrlParts := strings.Split(formerUrl, "/")

			if formerUrlParts[1] == "api" {
				cs.ClientSet = formerUrl
				return nil
			}

			group := formerUrlParts[len(formerUrlParts)-3]
			preferVersion := apiPrefer[group]
			formerUrlParts[len(formerUrlParts)-2] = preferVersion
			cs.ClientSet = strings.Join(formerUrlParts, "/")
			return nil
		}
	}

	return errors.New("url not found, please check k8s api docs")
}

//AddVersionIntoBody AddVersionIntoBody
func (cs *ClientSetter) AddVersionIntoBody() error {
	bodyJson := json.Get(*cs.BodyContent).GetInterface().(map[string]interface{})
	bodyJson["apiVersion"] = cs.ClientSet
	bodyByte, err := json.Marshal(bodyJson)
	if err != nil {
		return err
	}
	cs.BodyContent = &bodyByte
	return nil
}

//FormatURI FormatURI
func FormatURI(uri string) string {
	uriSplitParts := strings.Split(uri, "/")
	for index, uriSplitPart := range uriSplitParts {
		if index == 0 && len(uriSplitParts) >= 2 {
			if uriSplitPart == "namespaces" {
				uriSplitParts[1] = "{namespace}"
			} else {
				uriSplitParts[1] = "{name}"
			}
		}
		if index != 1 && index%2 != 0 {
			if uriSplitParts[index-1] == "proxy" {
				uriSplitParts[index] = "{path}"
			} else {
				uriSplitParts[index] = "{name}"
			}
		}
	}
	return strings.Join(uriSplitParts, "/")
}

//SplitUrlByVersion SplitUrlByVersion
func SplitUrlByVersion(url string) (string, string, error) {
	r, err := regexp.Compile("/v[0-9][a-z0-9]*/")
	if err != nil {
		return "", "", err
	}
	indexs := r.FindStringIndex(url)
	if len(indexs) < 1 {
		return "", "", errors.New("doesn't exist version part")
	}
	lastPosition := indexs[1]
	return url[0:lastPosition], url[lastPosition:], nil
}

//FetchAllUrl fetch all k8s api list from
func FetchAllUrl(version string) ([]string, error) {
	jsonFiles, err := FetchAllJsonFiles("metadata/*.json")

	if err != nil {
		return nil, err
	}
	for _, jsonFile := range jsonFiles {
		if jsonFile != "metadata/"+version+".json" {
			continue
		}
		urlsByVersion, err := ReadJsonFile(jsonFile, "paths")
		if err != nil {
			return nil, err
		}
		return urlsByVersion.Keys(), nil
	}
	return nil, fmt.Errorf("%s version is not supported yet", version)
}

//FetchAllJsonFiles FetchAllJsonFiles
func FetchAllJsonFiles(jsonPath string) ([]string, error) {
	files, err := filepath.Glob(jsonPath)
	if err != nil {
		return nil, err
	}
	return files, nil
}

//ReadJsonFile ReadJsonFile
func ReadJsonFile(jsonFile string, path string) (jsoniter.Any, error) {
	bytes, err := ioutil.ReadFile(jsonFile)
	if err != nil {
		return nil, err
	}
	// convert raw json content into []apiInfo
	apiList := json.Get(bytes, path)
	return apiList, nil
}

File: bcs-ui/backend/uniapps/network/views/charts/versions.py
# -*- coding: utf-8 -*-
"""
Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community
Edition) available.
Copyright (C) 2017-2021 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://opensource.org/licenses/MIT

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
"""
from rest_framework import viewsets
from rest_framework.renderers import BrowsableAPIRenderer
from rest_framework.response import Response

from backend.helm.helm.models import ChartVersion
from backend.uniapps.network import constants, serializers
from backend.uniapps.network.constants import K8S_LB_CHART_NAME, K8S_LB_NAMESPACE
from backend.uniapps.network.views.charts.releases import HelmReleaseMixin
from backend.utils.renderers import BKAPIRenderer


class K8SIngressControllerViewSet(viewsets.ViewSet, HelmReleaseMixin):
    renderer_classes = (BKAPIRenderer, BrowsableAPIRenderer)
    chart_name = K8S_LB_CHART_NAME
    namespace = K8S_LB_NAMESPACE
    public_repo_name = "public-repo"
    release_version_prefix = constants.RELEASE_VERSION_PREFIX

    def get_chart_versions(self, request, project_id):
        # 过滤公共仓库下面的lb chart名称
        chart_versions = (
            ChartVersion.objects.filter(
                name=self.chart_name,
                chart__repository__project_id=project_id,
                chart__repository__name=self.public_repo_name,
            )
            .order_by("-created")
            .values("version", "id")
        )
        # 获取release版本的version
        params = request.query_params
        cluster_id = params.get("cluster_id")
        # 查询是否release
        namespace = params.get("namespace") or self.namespace
        release = self.get_helm_release(cluster_id, name=self.chart_name, namespace=namespace)
        if not release:
            return Response(chart_versions)
        # id: -1, 表示此数据为组装数据，仅供前端展示匹配使用
        chart_versions = list(chart_versions)
        chart_versions.insert(
            0, {"version": f"{self.release_version_prefix} {release.get_current_version()}", "id": -1}
        )
        return Response(chart_versions)

    def get_version_detail(self, request, project_id):
        """获取指定版本chart信息，包含release的版本"""
        slz = serializers.ChartVersionSLZ(data=request.data)
        slz.is_valid(raise_exception=True)
        data = slz.validated_data
        version = data["version"]
        # 如果是release，查询release中对应的values信息
        if not version.startswith(self.release_version_prefix):
            chart_version = ChartVersion.objects.get(
                name=self.chart_name,
                version=version,
                chart__repository__project_id=project_id,
                chart__repository__name=self.public_repo_name,
            )
            version_detail = {"name": self.chart_name, "version": version, "files": chart_version.files}
            return Response(version_detail)
        # 获取release对应的values
        version_detail = {"name": self.chart_name, "version": version}
        namespace = data.get("namespace") or self.namespace
        cluster_id = data.get("cluster_id")
        release = self.get_helm_release(cluster_id, self.chart_name, namespace=namespace)
        if not release:
            return Response(version_detail)

        version_detail["files"] = release.release.chartVersionSnapshot.files
        # values 配置应该使用 release 的而不是特定版本 chart 默认的
        version_detail["files"][f"{self.chart_name}/values.yaml"] = release.release.valuefile

        return Response(version_detail)
